tests/tag_test.py

```diff
@@ -0,0 +1,169 @@
+experimental_tag_resolvable = "exp_w_2021_22"
+unknown_tag = "not_a_normal_format"
+mixed_case_tag = "MiXeD_CaSe_TaG"
+
+semver_tags = []
+for tt in [base_semver_tags, cycle_tags, rest_tags, both_tags]:
+    semver_tags.extend(tt)
+
+non_semver_tags = [
+    alias_tag,
+    experimental_tag_basic,
+    experimental_tag_resolvable,
+    unknown_tag,
+    mixed_case_tag,
+]
+
+tags = []
+for tt in [semver_tags, non_semver_tags]:
+    tags.extend(tt)
+
+# Make RubinTag objects from our tag list
+
+tagobjs = [RubinTag.from_tag(x, alias_tags=[alias_tag]) for x in tags]
+
+# Set up our expected tag types
+expected_semver_base_types = [
+    RubinTagType.RELEASE,
+    RubinTagType.RELEASE_CANDIDATE,
+    RubinTagType.WEEKLY,
+    RubinTagType.DAILY,
+]
+expected_semver_types = expected_semver_base_types * 4
+expected_non_semver_types = [
+    RubinTagType.ALIAS,
+    RubinTagType.EXPERIMENTAL,
+    RubinTagType.EXPERIMENTAL,
+    RubinTagType.UNKNOWN,
+    RubinTagType.UNKNOWN,
+]
+expected_tag_types = []
+for xt in [expected_semver_types, expected_non_semver_types]:
+    expected_tag_types.extend(xt)
+
+# Set up our expected semantic version strings
+
+expected_semver_base_strings = [
+    "21.0.1",
+    "22.0.0-rc1",
+    "2021.22.0",
+    "2021.5.27",
+]
+expected_semver_cycle_strings = [
+    f"{x}+{cycle_tag}" for x in expected_semver_base_strings
+]
+expected_semver_build_strings = [
+    f"{x}+{rest_tag}" for x in expected_semver_base_strings
+]
+expected_semver_both_strings = [
+    f"{x}+{cycle_tag}.{rest_tag}" for x in expected_semver_base_strings
+]
+expected_non_semver_strings = ["None"] * 5
+
+expected_semver = []
+for xs in [
+    expected_semver_base_strings,
+    expected_semver_cycle_strings,
+    expected_semver_build_strings,
+    expected_semver_both_strings,
+    expected_non_semver_strings,
+]:
+    expected_semver.extend(xs)
+
+# Set up our expected display names
+expected_base_semver_display_names = [
+    "Release r21.0.1",
+    "Release Candidate r22.0.0-rc1",
+    "Weekly 2021_22",
+    "Daily 2021_05_27",
+]
+
+cycle_names = [f"{x}_{cycle_tag}" for x in expected_base_semver_display_names]
+build_names = [f"{x}_{rest_tag}" for x in expected_base_semver_display_names]
+both_names = [
+    f"{x}_{cycle_tag}_{rest_tag}" for x in expected_base_semver_display_names
+]
+
+expected_non_semver_display_names = [
+    "Recommended",
+    "Experimental random",
+    "Experimental Weekly 2021_22",
+    "not_a_normal_format",
+    "MiXeD_CaSe_TaG",
+]
+
+expected_display_names = []
+for dn in [
+    expected_base_semver_display_names,
+    cycle_names,
+    build_names,
+    both_names,
+    expected_non_semver_display_names,
+]:
+    expected_display_names.extend(dn)
+
+expected_cycle = 20
+
+expected_cycles: List[Optional[int]] = []
+for t in [
+    [None] * 4,
+    [expected_cycle] * 4,
+    [None] * 4,
+    [expected_cycle] * 4,
+    [None] * 5,
+]:
+    expected_cycles.extend(t)  # type: ignore[arg-type]
```

Comment:
This is all very complicated and programmatic and it leaves me with no understanding of what the input was and what the output was expected to be. I think a clearer test for humans is to simply have a list of tuples where the first element of the tuple is the tag as a string and the second element of the tuple is the `RubinTag` object that we expect to get out of parsing that string. Then always call the parser with the same `alias_tags` setting (so you can include some aliases in the list).

In other words, I think the test should just be a long list of:

```
("index.docker.io/lsstsqre/sciplat-lab:r23_0_0_rc1_c0020.001_20210513",
    RubinTag(
        tag="r23_0_0_rc1_c0020.001_20210513",
        image_ref="index.docker.io/lsstsqre/sciplat-lab:r23_0_0_rc1_c0020.001_20210513",
        digest=None,
        image_type=RubinImageType.RELEASE_CANDIDATE,
        display_name="Release ...",
        semantic_version=VersionInfo("23.0.0rc1",
        cycle=20,
    )
)
```
I know it's tedious to write those all out (you may want to use this code as a one-off script to generate that test set), but it's incredibly helpful for humans when understanding and adding new tests to have the tests just say here is the input and here is the expected output.

Reply:
Done (using the RubinPartialTag class, to just exercise the parser)